File: plugins/npm/handler.py
#!/usr/bin/env python3
"""
npm domain handler for world.

Protocol: reads a JSON request from stdin, writes a JSON response to stdout.

Observes npm packages in the local project (node_modules) or globally (--global).
Target interpretation:
  - None / no target → list packages in nearest package.json project
  - "global"         → list globally installed packages
  - "<name>"         → show details for a specific package
"""

import json
import os
import subprocess
import sys


def observe(target):
    """Return structured npm state."""
    node_version = _run(["node", "--version"]).strip().lstrip("v")
    npm_version = _run(["npm", "--version"]).strip()

    is_global = target == "global"

    # Detect project context
    project = None
    pkg_json = _find_package_json()
    if pkg_json:
        try:
            with open(pkg_json) as f:
                project = json.load(f).get("name")
        except (json.JSONDecodeError, OSError):
            pass

    # Specific package lookup
    if target and target != "global":
        return _observe_package(target, node_version, npm_version, project)

    # List installed packages
    cmd = ["npm", "ls", "--json", "--depth=0"]
    if is_global:
        cmd.append("--global")

    installed = {}
    raw = _run(cmd, check=False)
    if raw:
        try:
            installed = json.loads(raw).get("dependencies", {})
        except json.JSONDecodeError:
            pass

    # Check outdated
    outdated_cmd = ["npm", "outdated", "--json"]
    if is_global:
        outdated_cmd.append("--global")
    outdated = {}
    raw_outdated = _run(outdated_cmd, check=False)
    if raw_outdated:
        try:
            outdated = json.loads(raw_outdated)
        except json.JSONDecodeError:
            pass

    packages = []
    for name, info in installed.items():
        current = info.get("version", "unknown")
        out_info = outdated.get(name, {})
        latest = out_info.get("latest")
        packages.append({
            "name": name,
            "version": current,
            "latest_version": latest,
            "outdated": name in outdated,
        })

    return {
        "details": {
            "packages": packages,
            "node_version": node_version,
            "npm_version": npm_version,
            "project": project if not is_global else None,
            "global": is_global,
        }
    }


def _observe_package(name, node_version, npm_version, project):
    """Observe a single package."""
    raw = _run(["npm", "ls", name, "--json", "--depth=0"], check=False)
    installed = False
    version = None
    if raw:
        try:
            deps = json.loads(raw).get("dependencies", {})
            if name in deps:
                installed = True
                version = deps[name].get("version")
        except json.JSONDecodeError:
            pass

    # Check latest from registry
    latest = None
    raw_view = _run(["npm", "view", name, "version", "--json"], check=False)
    if raw_view:
        try:
            latest = json.loads(raw_view)
        except json.JSONDecodeError:
            pass

    return {
        "details": {
            "packages": [{
                "name": name,
                "version": version,
                "latest_version": latest,
                "outdated": version != latest if version and latest else False,
            }],
            "node_version": node_version,
            "npm_version": npm_version,
            "project": project,
            "global": False,
        }
    }


def act(handler, target, params, dry_run):
    """Execute an npm action."""
    if handler == "install":
        version = (params or {}).get("version")
        pkg = f"{target}@{version}" if version else target
        if dry_run:
            return {"details": {"dry_run": True, "would_run": f"npm install {pkg}"}}
        result = subprocess.run(
            ["npm", "install", pkg],
            capture_output=True, text=True
        )
        if result.returncode != 0:
            return {"error": {"code": "install_failed", "message": result.stderr.strip()}}
        return {"details": {"installed": target, "output": result.stdout.strip()}}

    elif handler == "uninstall":
        if dry_run:
            return {"details": {"dry_run": True, "would_run": f"npm uninstall {target}"}}
        result = subprocess.run(
            ["npm", "uninstall", target],
            capture_output=True, text=True
        )
        if result.returncode != 0:
            return {"error": {"code": "uninstall_failed", "message": result.stderr.strip()}}
        return {"details": {"uninstalled": target}}

    elif handler == "pin_version":
        version = (params or {}).get("version")
        if not version:
            return {"error": {"code": "missing_param", "message": "version= required for set"}}
        pkg = f"{target}@{version}"
        if dry_run:
            return {"details": {"dry_run": True, "would_run": f"npm install {pkg}"}}
        result = subprocess.run(
            ["npm", "install", pkg],
            capture_output=True, text=True
        )
        if result.returncode != 0:
            return {"error": {"code": "pin_failed", "message": result.stderr.strip()}}
        return {"details": {"pinned": target, "version": version}}

    elif handler == "update_all":
        if dry_run:
            return {"details": {"dry_run": True, "would_run": "npm update"}}
        result = subprocess.run(
            ["npm", "update"],
            capture_output=True, text=True
        )
        if result.returncode != 0:
            return {"error": {"code": "update_failed", "message": result.stderr.strip()}}
        return {"details": {"updated": True, "output": result.stdout.strip()}}

    else:
        return {"error": {"code": "unknown_handler", "message": f"Unknown handler: {handler}"}}


def _find_package_json():
    """Walk up from cwd to find nearest package.json."""
    d = os.getcwd()
    while True:
        candidate = os.path.join(d, "package.json")
        if os.path.isfile(candidate):
            return candidate
        parent = os.path.dirname(d)
        if parent == d:
            return None
        d = parent


def _run(cmd, check=True):
    """Run a command, return stdout or empty string."""
    try:
        result = subprocess.run(
            cmd, capture_output=True, text=True, timeout=30
        )
        if check and result.returncode != 0:
            return ""
        return result.stdout
    except (subprocess.TimeoutExpired, FileNotFoundError):
        return ""


def main():
    request = json.load(sys.stdin)
    command = request.get("command")

    if command == "observe":
        response = observe(request.get("target"))
    elif command == "act":
        response = act(
            request["handler"],
            request.get("target"),
            request.get("params"),
            request.get("dry_run", False),
        )
    else:
        response = {"error": {"code": "unknown_command", "message": f"Unknown: {command}"}}

    json.dump(response, sys.stdout, indent=2)


if __name__ == "__main__":
    main()

File: plugins/npm/handler.js
#!/usr/bin/env node
/**
 * npm domain handler for world.
 *
 * Protocol: reads a JSON request from stdin, writes a JSON response to stdout.
 *
 * Target interpretation:
 *   - null / no target → list packages in nearest package.json project
 *   - "global"         → list globally installed packages
 *   - "<name>"         → show details for a specific package
 */

const { execFileSync } = require("child_process");
const { readFileSync, existsSync } = require("fs");
const { join, dirname } = require("path");

// ─── Helpers ──────────────────────────────────────────────────────────────────

function run(cmd, args, { check = true } = {}) {
  try {
    return execFileSync(cmd, args, {
      encoding: "utf8",
      timeout: 30000,
      stdio: ["pipe", "pipe", "pipe"],
    });
  } catch (e) {
    if (check) return "";
    return e.stdout || "";
  }
}

function jsonRun(cmd, args, opts) {
  const raw = run(cmd, args, opts);
  if (!raw) return null;
  try {
    return JSON.parse(raw);
  } catch {
    return null;
  }
}

function findPackageJson() {
  let d = process.cwd();
  while (true) {
    const candidate = join(d, "package.json");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(d);
    if (parent === d) return null;
    d = parent;
  }
}

// ─── Observe ──────────────────────────────────────────────────────────────────

function observe(target) {
  const nodeVersion = run("node", ["--version"]).trim().replace(/^v/, "");
  const npmVersion = run("npm", ["--version"]).trim();

  const isGlobal = target === "global";

  // Detect project context
  let project = null;
  const pkgJson = findPackageJson();
  if (pkgJson) {
    try {
      project = JSON.parse(readFileSync(pkgJson, "utf8")).name || null;
    } catch {}
  }

  // Single package lookup
  if (target && target !== "global") {
    return observePackage(target, nodeVersion, npmVersion, project);
  }

  // List installed
  const lsArgs = ["ls", "--json", "--depth=0"];
  if (isGlobal) lsArgs.push("--global");
  const lsData = jsonRun("npm", lsArgs, { check: false });
  const deps = (lsData && lsData.dependencies) || {};

  // Check outdated
  const outdatedArgs = ["outdated", "--json"];
  if (isGlobal) outdatedArgs.push("--global");
  const outdated = jsonRun("npm", outdatedArgs, { check: false }) || {};

  const packages = Object.entries(deps).map(([name, info]) => ({
    name,
    version: info.version || "unknown",
    latest_version: outdated[name] ? outdated[name].latest : null,
    outdated: name in outdated,
  }));

  return {
    details: {
      packages,
      node_version: nodeVersion,
      npm_version: npmVersion,
      project: isGlobal ? null : project,
      global: isGlobal,
    },
  };
}

function observePackage(name, nodeVersion, npmVersion, project) {
  // Local install check
  const lsData = jsonRun("npm", ["ls", name, "--json", "--depth=0"], {
    check: false,
  });
  const deps = (lsData && lsData.dependencies) || {};
  const version = deps[name] ? deps[name].version : null;

  // Registry latest
  const latest = jsonRun("npm", ["view", name, "version", "--json"], {
    check: false,
  });

  return {
    details: {
      packages: [
        {
          name,
          version,
          latest_version: typeof latest === "string" ? latest : null,
          outdated: version && latest ? version !== latest : false,
        },
      ],
      node_version: nodeVersion,
      npm_version: npmVersion,
      project,
      global: false,
    },
  };
}

// ─── Act ──────────────────────────────────────────────────────────────────────

function act(handler, target, params, dryRun) {
  switch (handler) {
    case "install": {
      const version = (params || {}).version;
      const pkg = version ? `${target}@${version}` : target;
      if (dryRun)
        return { details: { dry_run: true, would_run: `npm install ${pkg}` } };
      try {
        const out = execFileSync("npm", ["install", pkg], {
          encoding: "utf8",
          stdio: ["pipe", "pipe", "pipe"],
        });
        return { details: { installed: target, output: out.trim() } };
      } catch (e) {
        return {
          error: {
            code: "install_failed",
            message: (e.stderr || e.message).trim(),
          },
        };
      }
    }

    case "uninstall": {
      if (dryRun)
        return {
          details: { dry_run: true, would_run: `npm uninstall ${target}` },
        };
      try {
        execFileSync("npm", ["uninstall", target], {
          encoding: "utf8",
          stdio: ["pipe", "pipe", "pipe"],
        });
        return { details: { uninstalled: target } };
      } catch (e) {
        return {
          error: {
            code: "uninstall_failed",
            message: (e.stderr || e.message).trim(),
          },
        };
      }
    }

    case "pin_version": {
      const version = (params || {}).version;
      if (!version)
        return {
          error: { code: "missing_param", message: "version= required for set" },
        };
      const pkg = `${target}@${version}`;
      if (dryRun)
        return { details: { dry_run: true, would_run: `npm install ${pkg}` } };
      try {
        execFileSync("npm", ["install", pkg], {
          encoding: "utf8",
          stdio: ["pipe", "pipe", "pipe"],
        });
        return { details: { pinned: target, version } };
      } catch (e) {
        return {
          error: {
            code: "pin_failed",
            message: (e.stderr || e.message).trim(),
          },
        };
      }
    }

    case "update_all": {
      if (dryRun)
        return { details: { dry_run: true, would_run: "npm update" } };
      try {
        const out = execFileSync("npm", ["update"], {
          encoding: "utf8",
          stdio: ["pipe", "pipe", "pipe"],
        });
        return { details: { updated: true, output: out.trim() } };
      } catch (e) {
        return {
          error: {
            code: "update_failed",
            message: (e.stderr || e.message).trim(),
          },
        };
      }
    }

    default:
      return {
        error: {
          code: "unknown_handler",
          message: `Unknown handler: ${handler}`,
        },
      };
  }
}

// ─── Main ─────────────────────────────────────────────────────────────────────

let input = "";
process.stdin.setEncoding("utf8");
process.stdin.on("data", (chunk) => (input += chunk));
process.stdin.on("end", () => {
  const request = JSON.parse(input);
  let response;

  if (request.command === "observe") {
    response = observe(request.target);
  } else if (request.command === "act") {
    response = act(
      request.handler,
      request.target,
      request.params,
      request.dry_run || false
    );
  } else {
    response = {
      error: {
        code: "unknown_command",
        message: `Unknown: ${request.command}`,
      },
    };
  }

  process.stdout.write(JSON.stringify(response, null, 2));
});
